source/mongodb-handshake/handshake.rst

```diff
@@ -224,6 +224,107 @@ Example::
+
+:since: 4.4
+
+The ``isMaster`` handshake supports a new argument, ``speculativeAuthenticate``,
+provided as a BSON object. Clients specifying this argument to ``isMaster`` will
+speculatively include the first command of an authentication handshake.
+This command may be provided to the server in parallel with any standard request for
+supported authentication mechanisms (i.e. ``saslSupportedMechs``). This would permit
+clients to merge their first message with their ``isMaster`` request, and
+receive the authentication reply with the ``isMaster`` reply.
+
+When the mechanism is ``MONGODB-X509``, `speculativeAuthenticate` has the following
+structure::
+
+    {
+        isMaster: 1,
+        speculativeAuthenticate: {
+            authenticate: 1,
+            mechanism: "MONGODB-X509"
+        }
+    }
+
+When the mechanism is ``SCRAM-SHA-1`` or ``SCRAM-SHA-256``, `speculativeAuthenticate`
+has the following structure::
+
+    {
+        isMaster: 1,
+        speculativeAuthenticate: {
+            saslStart: 1,
+            mechanism: "<string>",   /* "SCRAM-SHA-1" or "SCRAM-SHA-256" */
+            db: "<authentication database>",
+            payload: BinData(...),
+            options: {
+                skipEmptyExchange: true         /* REQUIRED */
+            }
+        }
+    }  
+
+
+speculativeAuthenticate.mechanism
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required.
+
+This value indicates which authentication mechanism to use with the credential.
+Currently, the only mechanism supported when ``speculativeAuthenticate.authenticate``
+is present is ``MONGODB-X509``.
+
+speculativeAuthenticate.db
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required.
+
+This value indicates the name of the database used for authentication. 
+The database used when ``speculativeAuthenticate.authenticate`` is present is
+``$external``. When ``speculativeAuthenticate.saslStart`` is present, the database
+name defaults to the database name provided in the connection string or ``admin``.
+
+speculativeAuthenticate.payload
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required when ``speculativeAuthenticate.saslStart`` is present.
+
+A sequence of bytes or a base64 encoded string (depending on input) to pass
+into the SASL library to transition the state machine.
+
+speculativeAuthenticate.options.skipEmptyExchange
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+This value is required when ``speculativeAuthenticate.saslStart`` is present.
+
+This value notifies newer servers to shorten the authentication conversation by
+one round trip. Older servers will ignore this field.
+
+If the ``isMaster`` command with a ``speculativeAuthenticate`` argument succeeds,
+the client should proceed with the next step of the exchange. If the ``isMaster``
+response does not include a ``speculativeAuthenticate`` reply and the ``ok`` field
+in the ``isMaster`` response is set to 1, drivers MUST authenticate using the standard
+authentication handshake.
+
```

Comment:
I think we should specify what the `speculativeAuthenticate` response looks like.  In particular, is it a full embedded command response with ok, errmsg, code, etc?

Reply:
I think this would help.

Reply:
If it has no ok field, how do clients know if authentication has failed? Does the ismaster reply itself have ok: 0 at the top level? I wouldn't think so.

Reply:
If authentication fails when sending `speculativeAuthenticate`, the `isMaster` response will not contain a `speculativeAuthenticate` reply. Drivers should use the standard authentication handshake in that case. The top-level `ok` field is not affected by a speculative authentication failure.

Reply:
@sgolemon, this seems odd.  It means that every failed authentication attempt will be tried twice before the driver reports the failure, since the driver can't distinguish the failure of the speculative one from one where the server doesn't support speculation.

Can you comment?

Reply:
@jstewart-mongo is correct.  The problem with reporting failure during speculation is *why* the failure would occur.

For SCRAM methods, a first step failure has nothing to do with incorrect password (since verification hash isn't sent till step two).  The most probable failure here (apart from unknown user because someone is trying to dictionary our auth db) is "incorrect mechanism (e.g. using SCRAM-SHA-256 when user is only configured for SCRAM-SHA-1".  For this, saslSupportedMechs is going to tell you the correct mechanism to use and a retry is appropriate.  If we returned an explicit failure, it would be easy (and understandable) for a client to (incorrectly) abort further authentication attempts without retrying using the saslSupportedMechs hint.

Note that this first attempt is "free" from a network point of view compared to existing authentication attempts, so the fallback to saslSupportedMechs is no worse than present.

For X509, we could have returned failure more eagerly, but that would have been inconsistent with SASL, so we kept the failure case the same.  Of course, with X509 the actual validation was performed during TLS establishment, asking for authentication here is just a formality.

Reply:
Thank you, Sara!  That makes a lot of sense now.